workspaces/ui-v2/src/optic-components/hooks/edit/Contributions.tsx

```diff
@@ -27,58 +26,101 @@ interface IContribution {
 
 type ContributionEditingStoreProps = {
   initialIsEditingState?: boolean;
-  children?: any;
 };
 
-export const ContributionEditingStore = (
-  props: ContributionEditingStoreProps
+export const useValueWithStagedContributions = (
+  id: string,
+  contributionKey: string,
+  initialValue: string
 ) => {
-  //@TODO: replace with spectacle
-  /////////////////////////////////////////
+  const { stagePendingContribution } = useContributionEditing();
+
+  const [value, setValue] = useState<string>(initialValue || '');
+  useEffect(() => {
+    stagePendingContribution(id, contributionKey, value, initialValue);
+  }, [id, contributionKey, value, stagePendingContribution, initialValue]);
+
+  return {
+    value,
+    setValue,
+  };
+};
 
+export const ContributionEditingStore: FC<ContributionEditingStoreProps> = (
+  props
+) => {
+  const spectacleMutator = useSpectacleCommand();
   const [isEditing, setIsEditing] = useState(
     props.initialIsEditingState || false
   );
-  const [pendingContributions, setPendingContributions]: [
-    IContribution[],
-    any
-  ] = useState([]);
 
-  const stagePendingContribution = (
-    id: string,
-    contributionKey: string,
-    value: string
-  ) => {
-    setPendingContributions((current: IContribution[]) => {
-      const items = [...current];
-      // remove previous changes for same id/key
-      return [
-        ...items.filter(
-          (previous) =>
-            previous.id !== id && previous.contributionKey !== contributionKey
-        ),
-        {
-          id,
-          contributionKey,
-          value,
-        },
-      ];
-    });
-  };
+  const [pendingContributions, setPendingContributions] = useState<
+    IContribution[]
+  >([]);
+
+  const stagePendingContribution = useCallback(
+    (
+      newId: string,
+      newContributionKey: string,
+      newValue: string,
+      initialValue: string
+    ) => {
+      const newContribution =
+        newValue !== initialValue
```

Comment:
So here's an interesting bug. initialValue is set outside of this store, but the Contributions store mutates the value - what happens when we send a new contribution is that our UI is out of sync from spectacle (and we'll need to trigger a refetch) otherwise we have a stale initialValue (which causes a very minor bug when you edit, and reedit the same field).

How we fix this is either:
- We have a global store that stores the state of the app so that data + updates are in a single place (this could be something like redux, xstate for the entire app or apollo)
- Force a rerender (since the useEffect hook can't tell that data is stale)

Ultimately this is a pretty minor issue though but I think having a global store does simplify + reduce these stores of issues